### PR TITLE
New SMT2 backend: arithmetic operators implementation for fixed size bit vectors

### PR DESCRIPTION
**Pull request by NlightNFotis**
This adds implementations (and tests) for the binary arithmetic operators (and their conversion) to SMT2 data structures.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by TGWDB — APPROVED** [review]
Looks good. It would be nice to have unit tests of the validation of bitvector width being the same for the binary operations to ensure this is properly caught.

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp [hunk 1]
```diff
@@ -686,12 +686,6 @@ smt_termt convert_expr_to_smt(const exprt &expr)
   {
     return convert_expr_to_smt(*float_not_equal);
   }
-  if(
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/convert_expr_to_smt.cpp]: 🛠️ Could you add an explanation of why this would have incorrectly called `binary_relation_exprt` previously to the commit message body? As in - "The casting operations for `binary_relation_exprt` incorrectly allow casting to this type for expressions which do not derive from it".

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp [hunk 2]
```diff
@@ -325,8 +325,12 @@ convert_expr_to_smt(const binary_relation_exprt &binary_relation)
 
 static smt_termt convert_expr_to_smt(const plus_exprt &plus)
 {
-  UNIMPLEMENTED_FEATURE(
-    "Generation of SMT formula for plus expression: " + plus.pretty());
+  for(const auto &op : plus.operands())
+  {
+    PRECONDITION(can_cast_type<bitvector_typet>(op.type()));
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/convert_expr_to_smt.cpp]: ⛏️ The way this is implemented leaves no place where the control flow "falls through" for non-bitvector types. This means that we have to refactor this implementation when we add extra support later. For example, the old `smt2_convt::convert_plus` also supports conversion of `ID_rational`, `ID_integer`, `ID_pointer` and `ID_vector` in addition to bitvectors. So we can expect to be adding these later.

File: src/solvers/smt2_incremental/smt_bit_vector_theory.cpp [hunk 3]
```diff
@@ -42,3 +42,29 @@ static void validate_bit_vector_predicate_arguments(
...
+{
+  return "bvadd";
+}
+
+smt_sortt smt_bit_vector_theoryt::additiont::return_sort(
+  const smt_termt &lhs,
+  const smt_termt &rhs)
+{
+  // For now, make sure that they have the same bit-width
+  additiont::validate(lhs, rhs);
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/smt_bit_vector_theory.cpp]: 🚫 The single caller of `return_sort` already calls `validate`. Adding this here means that it gets called twice. Remove this line please.

File: src/solvers/smt2_incremental/smt_bit_vector_theory.cpp [hunk 4]
```diff
@@ -42,3 +42,29 @@ static void validate_bit_vector_predicate_arguments(
...
+  const auto width = lhs.get_sort().cast<smt_bit_vector_sortt>()->bit_width();
+  return smt_bit_vector_sortt{width};
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/smt_bit_vector_theory.cpp]: 🚫  Extracting the `bit_width` from the existing `bit_vectory_sortt`, just to construct a new sort which is identical to the original sort is doing unnecessary work and is preventing `irept` sharing from doing its thing. Just use `return lhs.get_sort();`

File: src/solvers/smt2_incremental/smt_bit_vector_theory.cpp [hunk 5]
```diff
@@ -42,3 +42,29 @@ static void validate_bit_vector_predicate_arguments(
...
+}
+
+void smt_bit_vector_theoryt::additiont::validate(
+  const smt_termt &lhs,
+  const smt_termt &rhs)
+{
+  validate_bit_vector_predicate_arguments(lhs, rhs);
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/smt_bit_vector_theory.cpp]: 🚫 Calling a function named `validate_bit_vector_predicate_arguments` for validation of something which is not a predicate is highly misleading. I have no problem with you reusing the function, but it needs renaming in order to ensure the code remains readable.

File: src/solvers/smt2_incremental/smt_bit_vector_theory.h [hunk 6]
```diff
@@ -20,6 +20,14 @@ class smt_bit_vector_theoryt
...
 #include "smt_bit_vector_theory.def"
 #undef SMT_BITVECTOR_THEORY_PREDICATE
+
+  struct additiont final
+  {
+    static const char *identifier();
+    static smt_sortt return_sort(const smt_termt &lhs, const smt_termt &rhs);
+    static void validate(const smt_termt &lhs, const smt_termt &rhs);
+  };
+  static const smt_function_application_termt::factoryt<additiont> addition;
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/smt_bit_vector_theory.h]: 🚫 Rename `addition` to `add` please.

File: unit/solvers/smt2_incremental/convert_expr_to_smt.cpp [hunk 7]
```diff
@@ -239,3 +239,58 @@ TEST_CASE(
...
+  {
+    const auto constructed_term =
+      convert_expr_to_smt(plus_exprt{one_bvint, two_bvint});
+    const auto expected_term =
+      smt_bit_vector_theoryt::addition(smt_term_one, smt_term_two);
+    CHECK(constructed_term == expected_term);
+  }
+
+  SECTION(
+    "Addition of four counstant size bit-vectors - testing multi-ary handling "
```
Comment (thomasspriggs) [inline unit/solvers/smt2_incremental/convert_expr_to_smt.cpp]: ⛏️ `counstant` is not an (English) word.

File: unit/solvers/smt2_incremental/convert_expr_to_smt.cpp [hunk 8]
```diff
@@ -239,3 +239,58 @@ TEST_CASE(
...
+    CHECK(constructed_term == expected_term);
+  }
+
+  SECTION(
+    "Ensure that addition node conversion fails if the operands are not "
+    "bit-vector based")
+  {
+    const cbmc_invariants_should_throwt invariants_throw;
+    REQUIRE_THROWS(
+      convert_expr_to_smt(plus_exprt{true_exprt{}, false_exprt{}}));
```
Comment (thomasspriggs) [inline unit/solvers/smt2_incremental/convert_expr_to_smt.cpp:294]: 💡 It would also be good to test that malformed instances of `plus_exprt` with 0 or 1 operands fail an invariant.

File: unit/solvers/smt2_incremental/smt_bit_vector_theory.cpp [hunk 9]
```diff
@@ -166,3 +166,24 @@ TEST_CASE("SMT bit vector predicates", "[core][smt2_incremental]")
...
+    const auto function_application =
+      smt_bit_vector_theoryt::addition(two, three);
+    REQUIRE(
+      function_application.function_identifier() ==
+      smt_identifier_termt("bvadd", smt_bit_vector_sortt{8}));
+    REQUIRE(function_application.get_sort() == smt_bit_vector_sortt{8});
+    REQUIRE(function_application.arguments().size() == 2);
+    REQUIRE(function_application.arguments()[0].get() == two);
+    REQUIRE(function_application.arguments()[1].get() == three);
+  }
```
Comment (thomasspriggs) [inline unit/solvers/smt2_incremental/smt_bit_vector_theory.cpp:326]: ⛏️ We should test that calling `smt_bit_vector_theoryt::add` with bit_vectors of mismatched width or on bool sorted terms fails an invariant.

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp [hunk 10]
```diff
@@ -325,8 +325,12 @@ convert_expr_to_smt(const binary_relation_exprt &binary_relation)
 
 static smt_termt convert_expr_to_smt(const plus_exprt &plus)
 {
-  UNIMPLEMENTED_FEATURE(
-    "Generation of SMT formula for plus expression: " + plus.pretty());
+  for(const auto &op : plus.operands())
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/convert_expr_to_smt.cpp]: ⛏️ `op` should be `operand` given the coding standards guideline to "avoid abbreviations". `op` is a particularly ambiguous abbreviation given that you need to read the context to know if it is an "operator" / "operation" or "operand".

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp (hunk 1 above)
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/convert_expr_to_smt.cpp]: ⛏️ Can you keep commit titles short enough that they display in their entirety in the github UI without having to expand them?

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp [hunk 11]
```diff
@@ -325,8 +325,12 @@ convert_expr_to_smt(const binary_relation_exprt &binary_relation)
 
 static smt_termt convert_expr_to_smt(const plus_exprt &plus)
 {
-  UNIMPLEMENTED_FEATURE(
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/convert_expr_to_smt.cpp:329]: ⛏️ I would prefer it if commits could be titled such that the title encompasses the entirety of what they change. The fact that you effectively have "Also add conversion of `plus_exprt` to `smt_bit_vector_theoryt::addition`" in the message body is kind of a red flag which suggests that you have combined unrelated changes together (even if one change depends of the other).

**Review by thomasspriggs — CHANGES_REQUESTED** [review]
This an initial review of the first two commits titled - "Change order of casting and dispatching of binary_relation_exprt in `convert_expr_to_smt`" and "Add implementation of arithmetic operator bvadd." only, so that I can leave some initial comments. My review of the remaining commits is still to follow.

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp [hunk 12]
```diff
@@ -335,8 +335,10 @@ static smt_termt convert_expr_to_smt(const plus_exprt &plus)
 
 static smt_termt convert_expr_to_smt(const minus_exprt &minus)
 {
-  UNIMPLEMENTED_FEATURE(
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/convert_expr_to_smt.cpp:335]: ⛏️ This should still hit an UNIMPLEMENTED_FEATURE macro, when this function encounters types which are not yet implemented. This is because the implementation cannot be considered to be complete as it currently stands.

File: src/solvers/smt2_incremental/smt_bit_vector_theory.cpp [hunk 13]
```diff
@@ -68,3 +68,29 @@ void smt_bit_vector_theoryt::additiont::validate(
...
+{
+  return "bvsub";
+}
+
+smt_sortt smt_bit_vector_theoryt::subtractiont::return_sort(
+  const smt_termt &lhs,
+  const smt_termt &rhs)
+{
+  // For now, make sure that they have the same bit-width
+  subtractiont::validate(lhs, rhs);
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/smt_bit_vector_theory.cpp]: 🚫 See previous comment on `::validate` for the plus operation.

File: src/solvers/smt2_incremental/smt_bit_vector_theory.h [hunk 14]
```diff
@@ -28,6 +28,15 @@ class smt_bit_vector_theoryt
...
   static const smt_function_application_termt::factoryt<additiont> addition;
+
+  struct subtractiont final
+  {
+    static const char *identifier();
+    static smt_sortt return_sort(const smt_termt &lhs, const smt_termt &rhs);
+    static void validate(const smt_termt &lhs, const smt_termt &rhs);
+  };
+  static const smt_function_application_termt::factoryt<subtractiont>
+    subtraction;
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/smt_bit_vector_theory.h]: 🚫 Rename `subtraction` to `subtract`, please.

File: unit/solvers/smt2_incremental/smt_bit_vector_theory.cpp [hunk 15]
```diff
@@ -186,4 +186,17 @@ TEST_CASE(
...
+  {
+    const auto function_application =
+      smt_bit_vector_theoryt::subtraction(two, three);
+    REQUIRE(
+      function_application.function_identifier() ==
+      smt_identifier_termt("bvsub", smt_bit_vector_sortt{8}));
+    REQUIRE(function_application.get_sort() == smt_bit_vector_sortt{8});
+    REQUIRE(function_application.arguments().size() == 2);
+    REQUIRE(function_application.arguments()[0].get() == two);
+    REQUIRE(function_application.arguments()[1].get() == three);
```
Comment (thomasspriggs) [inline unit/solvers/smt2_incremental/smt_bit_vector_theory.cpp:207]: ⛏️ Tests for invariant violation on constructing with an `lhs` and `rhs` of mismatched widths and terms of bool sort please.

File: src/solvers/smt2_incremental/smt_bit_vector_theory.h [hunk 16]
```diff
@@ -20,6 +20,56 @@ class smt_bit_vector_theoryt
...
+
+  struct multiplicationt final
+  {
+    static const char *identifier();
+    static smt_sortt return_sort(const smt_termt &lhs, const smt_termt &rhs);
+    static void validate(const smt_termt &lhs, const smt_termt &rhs);
+  };
+  static const smt_function_application_termt::factoryt<multiplicationt>
+    multiplication;
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/smt_bit_vector_theory.h]: 🚫 Rename `multiplication` to `multiply`, please.

File: src/solvers/smt2_incremental/smt_bit_vector_theory.h [hunk 17]
```diff
@@ -20,6 +20,56 @@ class smt_bit_vector_theoryt
...
+
+  struct divisiont final
+  {
+    static const char *identifier();
+    static smt_sortt return_sort(const smt_termt &lhs, const smt_termt &rhs);
+    static void validate(const smt_termt &lhs, const smt_termt &rhs);
+  };
+  static const smt_function_application_termt::factoryt<divisiont> division;
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/smt_bit_vector_theory.h]: 🚫 Rename `division` to `divide`, please.

File: src/solvers/smt2_incremental/smt_bit_vector_theory.h [hunk 18]
```diff
@@ -20,6 +20,56 @@ class smt_bit_vector_theoryt
...
+  };
+  static const smt_function_application_termt::factoryt<remaindert> remainder;
+
+  struct negationt final
+  {
+    static const char *identifier();
+    static smt_sortt return_sort(const smt_termt &operand);
+    static void validate(const smt_termt &operand);
+  };
+  static const smt_function_application_termt::factoryt<negationt> negation;
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/smt_bit_vector_theory.h]: 🚫 Rename `negation` to `negate`, please.

File: src/solvers/smt2_incremental/smt_bit_vector_theory.cpp [hunk 19]
```diff
@@ -42,3 +42,152 @@ static void validate_bit_vector_predicate_arguments(
...
+  validate_bit_vector_predicate_arguments(lhs, rhs);
+}
+
+const smt_function_application_termt::factoryt<
+  smt_bit_vector_theoryt::additiont>
+  smt_bit_vector_theoryt::addition{};
+
+const char *smt_bit_vector_theoryt::subtractiont::identifier()
+{
+  return "bvsub";
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/smt_bit_vector_theory.cpp:233]: ❗ `bvsub` does not appear to be a valid operator in bit vector theory, according to this definition - https://smtlib.cs.uiowa.edu/theories-FixedSizeBitVectors.shtml . We should probably exclude it from our definition of the theory and convert minus expressions to a combination of plus and negation terms instead.

File: src/solvers/smt2_incremental/smt_bit_vector_theory.cpp [hunk 20]
```diff
@@ -42,3 +42,152 @@ static void validate_bit_vector_predicate_arguments(
...
+  validate_bit_vector_predicate_arguments(lhs, rhs);
+}
+
+const smt_function_application_termt::factoryt<
+  smt_bit_vector_theoryt::multiplicationt>
+  smt_bit_vector_theoryt::multiplication{};
+
+const char *smt_bit_vector_theoryt::divisiont::identifier()
+{
+  return "bvdiv";
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/smt_bit_vector_theory.cpp]: ❗ This should be `bvudiv`.

File: src/solvers/smt2_incremental/smt_bit_vector_theory.cpp [hunk 21]
```diff
@@ -42,3 +42,152 @@ static void validate_bit_vector_predicate_arguments(
...
+}
+
+smt_sortt
+smt_bit_vector_theoryt::negationt::return_sort(const smt_termt &operand)
+{
+  return operand.get_sort();
+}
+
+void smt_bit_vector_theoryt::negationt::validate(const smt_termt &operand)
+{
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/smt_bit_vector_theory.cpp:380]: 🚫 Missing validation. This should check that the operand has bitvector sort.

**Review by thomasspriggs — CHANGES_REQUESTED** [review]

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp (hunk 2 above)
Comment (NlightNFotis) [reply src/solvers/smt2_incremental/convert_expr_to_smt.cpp]: My rationale for this is that for now we do not support them (or more precisely, the operation is not supported for any type other than bitvectors) so I'm guarding for that type in the precondition. 

That check would be relaxed when needed (when support for the conversion of integers, rationals, etc are added).

Do you mean that you want a special message to that effect?

Comment (thomasspriggs) [reply src/solvers/smt2_incremental/convert_expr_to_smt.cpp]: I agree that this should have appropriate `INVARIANT`s. My argument is that this should be done based on an `UNIMPLMENTED` condition at the end of this function, rather than a `PRECONDITION` at the start. So the parts we have implemented are reached when conditions implemented are satisfied and we fall through to the `UNIMPLEMENTED` macro, if the given expression is not supported by any of the cases we have implemented.

My comment relates to how easy it will be to extend the implementation you have added, not really the ease of understanding it. Therefore I don't think additional messages would be helpful here.

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp [hunk 22]
```diff
@@ -154,9 +154,9 @@ static smt_termt convert_expr_to_smt(const bitnot_exprt &bitwise_not)
 
 static smt_termt convert_expr_to_smt(const unary_minus_exprt &unary_minus)
 {
-  UNIMPLEMENTED_FEATURE(
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/convert_expr_to_smt.cpp:157]: :no_entry_sign: This `UNIMPLEMENTED_FEATURE` should remain and be reachable, because conversion for types including `ID_rational`, `ID_integer`, `ID_real` and `ID_floatbv` are unimplemented. This comment also applies to lines of code with a :bird: comment.

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp [hunk 23]
```diff
@@ -154,9 +154,9 @@ static smt_termt convert_expr_to_smt(const bitnot_exprt &bitwise_not)
...
-    "Generation of SMT formula for unary minus expression: " +
-    unary_minus.pretty());
+  PRECONDITION(can_cast_type<bitvector_typet>(unary_minus.op().type()));
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/convert_expr_to_smt.cpp]: :no_entry_sign: General comment on the usage of `can_cast_type<bitvector_typet>` / `type_try_dynamic_cast<bitvector_typet>` - There are sub types of `bitvector_typet` for which the conversions which you have defined are not the correct behaviour and were not intended to be included in the current piece of work. These include `fixedbv_typet`, `floatbv_typet`, and `pointer_typet`. If expressions of these types are passed into the new expression to smt conversion then we must fail an `UNIMPLEMENTED` invariant rather than emitting the incorrect SMT conversion.

File: src/solvers/smt2_incremental/smt_bit_vector_theory.cpp (hunk 19 above)
Comment (thomasspriggs) [reply src/solvers/smt2_incremental/smt_bit_vector_theory.cpp:233]: Please disregard this comment, as `bvsub` is defined as part of `logic QF_BV`, rather than as part of `theory FixedSizeBitVectors`

File: src/solvers/smt2_incremental/smt_bit_vector_theory.cpp (hunk 20 above)
Comment (thomasspriggs) [reply src/solvers/smt2_incremental/smt_bit_vector_theory.cpp]: or `bvsdiv` as appropriate.

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp [hunk 24]
```diff
@@ -325,20 +325,28 @@ convert_expr_to_smt(const binary_relation_exprt &binary_relation)
...
+  PRECONDITION(can_cast_type<bitvector_typet>(minus.lhs().type()));
+  PRECONDITION(can_cast_type<bitvector_typet>(minus.rhs().type()));
+  return smt_bit_vector_theoryt::subtraction(
+    convert_expr_to_smt(minus.lhs()), convert_expr_to_smt(minus.rhs()));
 }
 
 static smt_termt convert_expr_to_smt(const div_exprt &divide)
 {
-  UNIMPLEMENTED_FEATURE(
-    "Generation of SMT formula for divide expression: " + divide.pretty());
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/convert_expr_to_smt.cpp:342]: :bird: See previous comment on unimplemented features.

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp [hunk 25]
```diff
@@ -367,8 +377,12 @@ convert_expr_to_smt(const euclidean_mod_exprt &euclidean_modulo)
 
 static smt_termt convert_expr_to_smt(const mult_exprt &multiply)
 {
-  UNIMPLEMENTED_FEATURE(
-    "Generation of SMT formula for multiply expression: " + multiply.pretty());
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/convert_expr_to_smt.cpp:372]: :bird: See previous comment on unimplemented features.

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp [hunk 26]
```diff
@@ -352,9 +360,11 @@ static smt_termt convert_expr_to_smt(const ieee_float_op_exprt &float_operation)
 
 static smt_termt convert_expr_to_smt(const mod_exprt &truncation_modulo)
 {
-  UNIMPLEMENTED_FEATURE(
-    "Generation of SMT formula for truncation modulo expression: " +
-    truncation_modulo.pretty());
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/convert_expr_to_smt.cpp:358]: :pick: We expect this operator to be applied only to (signed/unsigned) integer bit vectors. It might be good to document this expectation using a `UNEXPECTEDCASE` macro, similar to how `smt2_convt::convert_mod` does this.

File: src/solvers/smt2_incremental/smt_bit_vector_theory.h [hunk 27]
```diff
@@ -20,6 +20,56 @@ class smt_bit_vector_theoryt
...
+
+  struct divisiont final
+  {
+    static const char *identifier();
+    static smt_sortt return_sort(const smt_termt &lhs, const smt_termt &rhs);
+    static void validate(const smt_termt &lhs, const smt_termt &rhs);
+  };
+  static const smt_function_application_termt::factoryt<divisiont> division;
+
+  struct remaindert final
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/smt_bit_vector_theory.h]: :no_entry_sign: We actually need 2 variations of this operation because there is the unsigned remainder `bvurem` and the signed remainder `bvsrem`.

File: src/solvers/smt2_incremental/smt_bit_vector_theory.h [hunk 28]
```diff
@@ -20,6 +20,56 @@ class smt_bit_vector_theoryt
...
+  struct multiplicationt final
+  {
+    static const char *identifier();
+    static smt_sortt return_sort(const smt_termt &lhs, const smt_termt &rhs);
+    static void validate(const smt_termt &lhs, const smt_termt &rhs);
+  };
+  static const smt_function_application_termt::factoryt<multiplicationt>
+    multiplication;
+
+  struct divisiont final
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/smt_bit_vector_theory.h]: :no_entry_sign: We actually need 2 variations of this operation because there is the unsigned divide `bvudiv` and the signed divide `bvsdiv`.

**Review by thomasspriggs — CHANGES_REQUESTED** [review]

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp (hunk 1 above)
Comment (thomasspriggs) [reply src/solvers/smt2_incremental/convert_expr_to_smt.cpp]: This is now resolved due to rebasing on develop.

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp [hunk 29]
```diff
@@ -326,8 +326,21 @@ static optionalt<smt_termt> try_relational_conversion(const exprt &expr)
...
 {
-  UNIMPLEMENTED_FEATURE(
-    "Generation of SMT formula for plus expression: " + plus.pretty());
+  if(std::all_of(
+       plus.operands().cbegin(),
+       plus.operands().cend(),
+       [](exprt operand)
+       { return can_cast_type<bitvector_typet>(operand.type()); }))
+  {
+    // if all the operands are bitvectors, call the bitvector addition constructor
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/convert_expr_to_smt.cpp]: :pick: I don't consider this comment to be useful. This is because it states what can already been seen from reading the code, not why or how it does it.

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp (hunk 2 above)
Comment (thomasspriggs) [reply src/solvers/smt2_incremental/convert_expr_to_smt.cpp]: Looks good to me.

File: src/solvers/smt2_incremental/smt_bit_vector_theory.cpp (hunk 5 above)
Comment (thomasspriggs) [reply src/solvers/smt2_incremental/smt_bit_vector_theory.cpp]: This still needs to be addressed.

File: src/solvers/smt2_incremental/smt_bit_vector_theory.cpp (hunk 4 above)
Comment (thomasspriggs) [reply src/solvers/smt2_incremental/smt_bit_vector_theory.cpp]: This is now resolved.

File: src/solvers/smt2_incremental/smt_bit_vector_theory.h (hunk 6 above)
Comment (thomasspriggs) [reply src/solvers/smt2_incremental/smt_bit_vector_theory.h]: Could you do the equivalent rename for the class as well please? `additiont` -> `addt`.

File: unit/solvers/smt2_incremental/convert_expr_to_smt.cpp (hunk 7 above)
Comment (thomasspriggs) [reply unit/solvers/smt2_incremental/convert_expr_to_smt.cpp]: Looks to be resolved.

File: unit/solvers/smt2_incremental/convert_expr_to_smt.cpp (hunk 8 above)
Comment (thomasspriggs) [reply unit/solvers/smt2_incremental/convert_expr_to_smt.cpp:294]: Could we get the test for 1 operand as well as 0 operands?

File: src/solvers/smt2_incremental/smt_bit_vector_theory.cpp (hunk 3 above)
Comment (thomasspriggs) [reply src/solvers/smt2_incremental/smt_bit_vector_theory.cpp]: Looks ok to me.

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp (hunk 10 above)
Comment (thomasspriggs) [reply src/solvers/smt2_incremental/convert_expr_to_smt.cpp]: Looks good to me.

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp (hunk 1 above)
Comment (thomasspriggs) [reply src/solvers/smt2_incremental/convert_expr_to_smt.cpp]: This looks to be resolved.

File: src/solvers/smt2_incremental/smt_bit_vector_theory.cpp (hunk 5 above)
Comment (thomasspriggs) [reply src/solvers/smt2_incremental/smt_bit_vector_theory.cpp]: This comment has now been resolved by the rename commit.

File: src/solvers/smt2_incremental/smt_bit_vector_theory.h (hunk 6 above)
Comment (thomasspriggs) [reply src/solvers/smt2_incremental/smt_bit_vector_theory.h]: This is now resolved.

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp [hunk 30]
```diff
@@ -366,10 +366,22 @@ static smt_termt convert_expr_to_smt(const div_exprt &divide)
     can_cast_type<integer_bitvector_typet>(divide.lhs().type()) &&
     can_cast_type<integer_bitvector_typet>(divide.rhs().type());
 
+  bool both_operands_unsigned =
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/convert_expr_to_smt.cpp]: :pick: `const` - Every variable which is not mutated should be declared `const`. I have picked this as an arbitrary example, but I think it is missing in a few places.

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp [hunk 31]
```diff
@@ -366,10 +366,22 @@ static smt_termt convert_expr_to_smt(const div_exprt &divide)
...
-      convert_expr_to_smt(divide.lhs()), convert_expr_to_smt(divide.rhs()));
+    if(both_operands_unsigned)
+    {
+      return smt_bit_vector_theoryt::unsigned_divide(
+        convert_expr_to_smt(divide.lhs()), convert_expr_to_smt(divide.rhs()));
+    }
+    else
+    {
+      return smt_bit_vector_theoryt::signed_divide(
+        convert_expr_to_smt(divide.lhs()), convert_expr_to_smt(divide.rhs()));
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/convert_expr_to_smt.cpp:392]: :bulb: I suggest introducing `lhs_term` and `rhs_term` variables above this if statement, so that the calls to `convert_expr_to_smt` are not duplicated in the two branches.

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp [hunk 32]
```diff
@@ -401,9 +401,24 @@ static smt_termt convert_expr_to_smt(const ieee_float_op_exprt &float_operation)
...
+
+  if(both_operands_bitvector)
+  {
+    return smt_bit_vector_theoryt::unsigned_remainder(
+      convert_expr_to_smt(truncation_modulo.lhs()),
+      convert_expr_to_smt(truncation_modulo.rhs()));
+  }
+  else
+  {
+    {
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/convert_expr_to_smt.cpp]: :pick: Superfluous brackets.

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp [hunk 33]
```diff
@@ -405,11 +405,24 @@ static smt_termt convert_expr_to_smt(const mod_exprt &truncation_modulo)
...
+      return smt_bit_vector_theoryt::unsigned_remainder(
+        convert_expr_to_smt(truncation_modulo.lhs()),
+        convert_expr_to_smt(truncation_modulo.rhs()));
+    }
+    else
+    {
+      return smt_bit_vector_theoryt::signed_remainder(
+        convert_expr_to_smt(truncation_modulo.lhs()),
+        convert_expr_to_smt(truncation_modulo.rhs()));
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/convert_expr_to_smt.cpp:433]: :bulb: Factor out the common `convert_expr_to_smt` sub-expressions, like I suggested with your implementation for the divide operations.

**Review by thomasspriggs — APPROVED** [review]
I am now approving as all of my :no_entry_sign: comments have been addressed.
